pkg/apis/workflow/v1alpha1/workflow_types.go

```diff
@@ -1004,6 +998,51 @@ func (in ResourcesDuration) IsZero() bool {
 	return len(in) == 0
 }
 
+type WorkflowConditions []*WorkflowCondition
+
+func (wc *WorkflowConditions) AddOrReplaceCondition(condition *WorkflowCondition) {
+	for index, wfCondition := range *wc {
+		if wfCondition.Type == condition.Type {
+			(*wc)[index] = condition
+			return
+		}
+	}
+	*wc = append(*wc, condition)
+}
+
+func (wc *WorkflowConditions) AddOrAppendCondition(condition *WorkflowCondition) {
+	for index, wfCondition := range *wc {
+		if wfCondition.Type == condition.Type {
+			(*wc)[index].Message += ", " + condition.Message
+			return
+		}
+	}
+	*wc = append(*wc, condition)
+}
+
+func (wc *WorkflowConditions) JoinConditions(conditions *WorkflowConditions) {
+	for _, condition := range *conditions {
+		wc.AddOrReplaceCondition(condition)
+	}
+}
+
+type WorkflowConditionType string
+
+const (
+	// WorkflowConditionCompleted is a signifies the workflow has completed
+	WorkflowConditionCompleted WorkflowConditionType = "Completed"
+	// WorkflowConditionSpecWarning is a warning on the current application spec
+	WorkflowConditionSpecWarning WorkflowConditionType = "SpecWarning"
+)
+
+type WorkflowCondition struct {
+	Type WorkflowConditionType `json:"type,omitempty" protobuf:"bytes,1,opt,name=type,casttype=WorkflowConditionType"`
```

Comment:
I think you don't want `omitempty`?
* What is casttype please?

Reply:
> I think you don't want omitempty?

Makes no difference since this field is read-only.

> What is casttype please?

Not sure, it was added by the codegen.